File: src/autograd/ParamManager.m
% The class ParamManager manages all optimizable parameters. We only have one instance of this class as global variable.
% Before the optimization starts, we need to simulate the constraint evaluation process and register the parameters.
% Each registered parameter can be a scalar or a row vector.
% When registering, this class returns an ID for that parameter (group). That ID will be stored by other entities.
% When the optimization starts, the evaluator will call this class for GVars indicating these variables.

% Known issues:
%  - The PackResults() function might be slow.

% Note:
%  - Mostly, we do not run the optimization process from a random initial point. However, the functionality of
%    generating a random point in the parameter space still remains, because we need to evaluate at random points
%    to see the sparsity structure of the matrix of gradients.

classdef ParamManager < handle
properties
  num_group
  num_input       % i.e., the number of optimizable parameters.
  group_startpos  % Array of starting positions. 1 by num_group.
  group_size
  lb              % Concatenated array of lower bounds. 1 by num_input.
  ub
  cur_x           % Current value of the parameters. 1 by num_input.
  initializer     % Initializer for the parameters. 1 by num_input cell-array.
  lin_A
  lin_b
  lin_Aeq
  lin_beq
end
methods
  % Constructor
  function obj = ParamManager()
    obj.num_group = 0;
    obj.num_input = 0;
    obj.group_startpos = [];
    obj.group_size = [];
    obj.lb = [];
    obj.ub = [];
    obj.initializer = {};
    obj.cur_x = [];
    obj.lin_A = sparse(0, 0);
    obj.lin_b = sparse(0, 0);
    obj.lin_Aeq = sparse(0, 0);
    obj.lin_beq = sparse(0, 0);
  end

  % Register a parameter. Return the ID of the group.
  % lb and ub can be omitted, in which case [-Inf, Inf] is applied.
  % lb and ub can be single element or match the size of the parameter.
  function group_id = Register(obj, param_len, lb, ub, initer)
    if nargin < 3
      lb = -Inf;
    end
    if nargin < 4
      ub = Inf;
    end
    if isscalar(lb)
      lb = repmat(lb, 1, param_len);
    end
    if isscalar(ub)
      ub = repmat(ub, 1, param_len);
    end
    obj.num_group = obj.num_group + 1;
    group_id = obj.num_group;
    obj.group_startpos(group_id) = obj.num_input + 1;
    obj.group_size(group_id) = param_len;
    obj.num_input = obj.num_input + param_len;
    obj.lb = [obj.lb, lb];
    obj.ub = [obj.ub, ub];

    % Set initializer
    % Supported types:
    %  - [l, r]: uniform distribution over [l, r]
    %  - a function handle (function pointer): the function should take the size of the parameter as input and return 1 by n row vector.
    if nargin < 5
      initer = [0, 0.01];
    end
    if isa(initer, 'function_handle')
      obj.initializer{group_id} = initer;
    else
      obj.initializer{group_id} = @(sz) initer(1) + (initer(2) - initer(1)) * rand(1, sz);
    end
  end

  % Specify the current values of the parameters.
  function SetValue(obj, x)
    obj.cur_x = min(max(x, obj.lb), obj.ub);
  end

  % Specify the current values for a single parameter group.
  function SetSingleParam(obj, group_id, value)
    obj.cur_x(obj.group_startpos(group_id) : obj.group_startpos(group_id) + obj.group_size(group_id) - 1) = value;
  end

  % Random initialization.
  function x = RandomInit(obj)
    x = zeros(1, obj.num_input);
    for i = 1 : obj.num_group
      x(obj.group_startpos(i) : obj.group_startpos(i) + obj.group_size(i) - 1) = obj.initializer{i}(obj.group_size(i));
    end
    obj.cur_x = x;
  end

  % Get GVar for a parameter.
  function gvar = GetParam(obj, group_id)
    cur_start_pos = obj.group_startpos(group_id);
    gvar = GVar(obj.num_input, obj.cur_x(cur_start_pos : cur_start_pos + obj.group_size(group_id) - 1), 'startpos', cur_start_pos);
  end

  % Pack results into desired format.
  function [y, dy] = PackResults(obj, y_gvar)
    % y_gvar is a 1-d cell array of GVars.
    % y is a column vector storing constraint values.
    % dy is (# of constraints) by (# of parameters) matrix storing the gradients of the constraints.
    % The matrices might be very large. The current implementation is not too slow, but maybe could be improved further.
    y = [];
    dy = sparse(0, obj.num_input);
    for i = 1 : length(y_gvar)
      num_new_rows = length(y_gvar{i}.value);
      y(end + 1 : end + num_new_rows, 1) = y_gvar{i}.value;
      dy = [dy; y_gvar{i}.grad'];
    end
  end

  % Linear Constraints
  % A_entries is a k by 2 cell array, each row is {group_id, coeff_matrix},
  %   where coeff_matrix is n by r (n is the size of the parameter group). r is the same for all rows in the cell array, and b is a row vector of length r.
  % It is supported to add constraints when some unrelated variables are not registered yet.
  function AddLinearConstraint(obj, A_entries, b)
    row_start = size(obj.lin_A, 1) + 1;
    r = length(b);
    obj.lin_A(row_start + r - 1, obj.num_input) = 0;  % resize the matrix
    obj.lin_b(row_start : row_start + r - 1, 1) = b';
    for i = 1 : size(A_entries, 1)
      group_id = A_entries{i, 1};
      coeff_matrix = A_entries{i, 2};
      obj.lin_A(row_start : row_start + r - 1, obj.group_startpos(group_id) : obj.group_startpos(group_id) + obj.group_size(group_id) - 1) = coeff_matrix';
    end
  end

  % Just the same as AddLinearConstraint, but Aeq & beq.
  function AddLinearConstraintEq(obj, A_entries, b)
    row_start = size(obj.lin_Aeq, 1) + 1;
    r = length(b);
    obj.lin_Aeq(row_start + r - 1, obj.num_input) = 0;  % resize the matrix
    obj.lin_beq(row_start : row_start + r - 1, 1) = b';
    for i = 1 : size(A_entries, 1)
      group_id = A_entries{i, 1};
      coeff_matrix = A_entries{i, 2};
      obj.lin_Aeq(row_start : row_start + r - 1, obj.group_startpos(group_id) : obj.group_startpos(group_id) + obj.group_size(group_id) - 1) = coeff_matrix';
    end
  end

  % Read the linear constraints.
  function [A, b, Aeq, beq] = GetLinearConstraints(obj)
    % The column number of A and Aeq might be smaller than num_input. Fix it.
    if size(obj.lin_A, 1) > 0 && size(obj.lin_A, 2) < obj.num_input
      obj.lin_A(end, obj.num_input) = 0;
    end
    if size(obj.lin_Aeq, 1) > 0 && size(obj.lin_Aeq, 2) < obj.num_input
      obj.lin_Aeq(end, obj.num_input) = 0;
    end
    % Set return values
    A = obj.lin_A;
    b = obj.lin_b;
    Aeq = obj.lin_Aeq;
    beq = obj.lin_beq;
    % Remove empty rows
    non_empty_rows = any(A, 2);
    A = A(non_empty_rows, :);
    b = b(non_empty_rows, :);
    non_empty_rows = any(Aeq, 2);
    Aeq = Aeq(non_empty_rows, :);
    beq = beq(non_empty_rows, :);
    % Save back
    obj.lin_A = A;
    obj.lin_b = b;
    obj.lin_Aeq = Aeq;
    obj.lin_beq = beq;
  end

  function Perturb(obj, perturb_size)
    obj.cur_x = obj.cur_x + perturb_size * randn(1, obj.num_input);
    obj.cur_x = min(max(obj.cur_x, obj.lb), obj.ub);
  end
end
end

File: fast_mm_python/src/autograd/ParamManager.py
"""
Parameter Manager for managing all optimizable parameters.

This class registers parameters, manages their bounds, and provides
access to GVar instances for constraint evaluation.
"""

import numpy as np
from scipy import sparse
from .GVar import GVar


# Global parameter manager instance
_param_manager = None


def get_param_manager():
    """Get the global parameter manager instance."""
    global _param_manager
    if _param_manager is None:
        _param_manager = ParamManager()
    return _param_manager


def set_param_manager(pm):
    """Set the global parameter manager instance."""
    global _param_manager
    _param_manager = pm


class ParamManager:
    """
    Manages all optimizable parameters for the optimization problem.
    
    Each parameter group can be a scalar or a vector. The manager
    assigns IDs to parameter groups and provides access through GVar instances.
    """
    
    def __init__(self):
        """Initialize empty parameter manager."""
        self.num_group = 0
        self.num_input = 0
        self.group_startpos = []  # Starting position of each group
        self.group_size = []      # Size of each group
        self.lb = np.array([])    # Lower bounds
        self.ub = np.array([])    # Upper bounds
        self.cur_x = np.array([]) # Current parameter values
        self.initializer = []     # Initializer functions for each group
        
        # Linear constraints: A @ x <= b
        self.lin_A = sparse.csr_matrix((0, 0), dtype=np.float64)
        self.lin_b = np.array([])
        
        # Linear equality constraints: Aeq @ x == beq
        self.lin_Aeq = sparse.csr_matrix((0, 0), dtype=np.float64)
        self.lin_beq = np.array([])
    
    def Register(self, param_len, lb=-np.inf, ub=np.inf, initializer=None):
        """
        Register a new parameter group.
        
        Args:
            param_len: Length of the parameter vector
            lb: Lower bound(s), scalar or array of length param_len
            ub: Upper bound(s), scalar or array of length param_len
            initializer: Initialization function or tuple (low, high) for uniform random
        
        Returns:
            group_id: ID of the registered parameter group
        """
        # Handle scalar bounds
        if np.isscalar(lb):
            lb = np.full(param_len, lb, dtype=np.float64)
        else:
            lb = np.array(lb, dtype=np.float64)
            
        if np.isscalar(ub):
            ub = np.full(param_len, ub, dtype=np.float64)
        else:
            ub = np.array(ub, dtype=np.float64)
        
        # Assign group ID
        group_id = self.num_group
        self.num_group += 1
        self.group_startpos.append(self.num_input)
        self.group_size.append(param_len)
        
        # Update bounds
        self.lb = np.concatenate([self.lb, lb])
        self.ub = np.concatenate([self.ub, ub])
        self.num_input += param_len
        
        # Set initializer
        if initializer is None:
            initializer = (0.0, 0.01)
        
        if callable(initializer):
            self.initializer.append(initializer)
        else:
            # Assume it's a tuple (low, high) for uniform random
            low, high = initializer
            self.initializer.append(lambda sz: np.random.uniform(low, high, sz))
        
        return group_id
    
    def SetValue(self, x):
        """
        Set current parameter values, clipped to bounds.
        
        Args:
            x: Parameter vector of length num_input
        """
        x = np.array(x, dtype=np.float64)
        self.cur_x = np.clip(x, self.lb, self.ub)
    
    def SetSingleParam(self, group_id, value):
        """
        Set the value for a single parameter group.
        
        Args:
            group_id: ID of the parameter group
            value: New value(s) for the group
        """
        start = self.group_startpos[group_id]
        end = start + self.group_size[group_id]
        expected_size = self.group_size[group_id]
        
        # Convert value to array and check size
        value_array = np.atleast_1d(value)
        if len(value_array) != expected_size:
            print(f"[WARN] SetSingleParam: group_id={group_id}, expected size={expected_size}, got size={len(value_array)}")
            # Pad or truncate as needed
            if len(value_array) < expected_size:
                value_array = np.pad(value_array, (0, expected_size - len(value_array)), constant_values=value_array[0])
            else:
                value_array = value_array[:expected_size]
        
        self.cur_x[start:end] = value_array
    
    def RandomInit(self):
        """
        Randomly initialize all parameters using their initializers.
        
        Returns:
            x: Randomly initialized parameter vector
        """
        x = np.zeros(self.num_input)
        for i in range(self.num_group):
            start = self.group_startpos[i]
            end = start + self.group_size[i]
            x[start:end] = self.initializer[i](self.group_size[i])
        
        self.cur_x = x
        return x
    
    def GetParam(self, group_id):
        """
        Get a GVar for a parameter group.
        
        Args:
            group_id: ID of the parameter group
        
        Returns:
            GVar instance with current values and appropriate gradients
        """
        start = self.group_startpos[group_id]
        end = start + self.group_size[group_id]
        param_size = self.group_size[group_id]
        
        # Ensure cur_x is initialized
        if len(self.cur_x) < self.num_input:
            # Initialize cur_x if not done yet
            self.cur_x = np.zeros(self.num_input)
        
        value = self.cur_x[start:end]
        
        # Double check we got the right size
        if len(value) != param_size:
            # This shouldn't happen, but if it does, return zeros
            value = np.zeros(param_size)
        
        return GVar(self.num_input, value, 'startpos', start)
    
    def PackResults(self, y_gvar_list):
        """
        Pack a list of GVars into constraint values and gradients.
        
        Args:
            y_gvar_list: List of GVar instances representing constraints
        
        Returns:
            y: Constraint values, shape (num_constraints,)
            dy: Constraint gradients, shape (num_constraints, num_input)
        """
        y_list = []
        dy_list = []
        
        for gvar in y_gvar_list:
            y_list.append(gvar.value)
            # Transpose: grad is (num_input, n) -> we want (n, num_input)
            dy_list.append(gvar.grad.T)
        
        # Concatenate all constraints
        if len(y_list) > 0:
            y = np.concatenate(y_list)
            dy = sparse.vstack(dy_list)
        else:
            y = np.array([])
            dy = sparse.csr_matrix((0, self.num_input))
        
        return y, dy
    
    def AddLinearConstraint(self, A_entries, b):
        """
        Add linear inequality constraints: A @ x <= b
        
        Args:
            A_entries: List of tuples (group_id, coeff_matrix)
                      where coeff_matrix has shape (n_constraints, group_size)
            b: Right-hand side vector of length n_constraints
        """
        b = np.atleast_1d(b)
        n_constraints = len(b)
        
        # Expand constraint matrix
        row_start = self.lin_A.shape[0]
        new_rows = n_constraints
        
        # Create new matrix with expanded size
        new_A = sparse.lil_matrix((row_start + new_rows, self.num_input))
        if row_start > 0:
            new_A[:row_start, :self.lin_A.shape[1]] = self.lin_A
        
        # Fill in new constraints
        for group_id, coeff_matrix in A_entries:
            start = self.group_startpos[group_id]
            end = start + self.group_size[group_id]
            new_A[row_start:row_start + new_rows, start:end] = coeff_matrix
        
        self.lin_A = new_A.tocsr()
        self.lin_b = np.concatenate([self.lin_b, b])
    
    def AddLinearConstraintEq(self, A_entries, b):
        """
        Add linear equality constraints: Aeq @ x == beq
        
        Args:
            A_entries: List of tuples (group_id, coeff_matrix)
            b: Right-hand side vector
        """
        b = np.atleast_1d(b)
        n_constraints = len(b)
        
        row_start = self.lin_Aeq.shape[0]
        new_rows = n_constraints
        
        new_Aeq = sparse.lil_matrix((row_start + new_rows, self.num_input))
        if row_start > 0:
            new_Aeq[:row_start, :self.lin_Aeq.shape[1]] = self.lin_Aeq
        
        for group_id, coeff_matrix in A_entries:
            start = self.group_startpos[group_id]
            end = start + self.group_size[group_id]
            new_Aeq[row_start:row_start + new_rows, start:end] = coeff_matrix
        
        self.lin_Aeq = new_Aeq.tocsr()
        self.lin_beq = np.concatenate([self.lin_beq, b])
    
    def GetLinearConstraints(self):
        """
        Get all linear constraints in standard form.
        
        Returns:
            A, b, Aeq, beq: Constraint matrices and vectors
        """
        # Ensure proper dimensions
        if self.lin_A.shape[0] > 0 and self.lin_A.shape[1] < self.num_input:
            new_A = sparse.lil_matrix((self.lin_A.shape[0], self.num_input))
            new_A[:, :self.lin_A.shape[1]] = self.lin_A
            self.lin_A = new_A.tocsr()
        
        if self.lin_Aeq.shape[0] > 0 and self.lin_Aeq.shape[1] < self.num_input:
            new_Aeq = sparse.lil_matrix((self.lin_Aeq.shape[0], self.num_input))
            new_Aeq[:, :self.lin_Aeq.shape[1]] = self.lin_Aeq
            self.lin_Aeq = new_Aeq.tocsr()
        
        # Remove empty rows
        if self.lin_A.shape[0] > 0:
            non_empty = np.array(self.lin_A.sum(axis=1)).flatten() != 0
            A = self.lin_A[non_empty, :]
            b = self.lin_b[non_empty]
        else:
            A = self.lin_A
            b = self.lin_b
        
        if self.lin_Aeq.shape[0] > 0:
            non_empty = np.array(self.lin_Aeq.sum(axis=1)).flatten() != 0
            Aeq = self.lin_Aeq[non_empty, :]
            beq = self.lin_beq[non_empty]
        else:
            Aeq = self.lin_Aeq
            beq = self.lin_beq
        
        # Save back cleaned versions
        self.lin_A = A
        self.lin_b = b
        self.lin_Aeq = Aeq
        self.lin_beq = beq
        
        return A, b, Aeq, beq
    
    def Perturb(self, perturb_size):
        """
        Add random perturbation to current parameters.
        
        Args:
            perturb_size: Standard deviation of Gaussian noise
        """
        noise = np.random.normal(0, perturb_size, self.num_input)
        self.cur_x = np.clip(self.cur_x + noise, self.lb, self.ub)

